### ISSUE TITLE
Failing to merge dicts with same supertype

### ISSUE THREAD

**Issue by Meemaw** [microsoft/pyright#3881]
**Describe the bug**
Trying to merge dicts using the `|` operator fails when for same supertype.

**Expected behavior**
This should work

**Screenshots or Code**
```python
class A:
    pass


class B(A):
    pass


overrides: dict[str, A] = {}

merged: dict[str, A] = {"a": B()} | overrides # this line fails.
```
If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**Additional context**
Pyright version 1.1.268

**Comment by erictraut** [comment]
The inferred type of the subexpression `{"a": B()}` is `dict[str, B]`. The operator `|` is implemented as a `__ror__` method override on the `dict` class, and it is declared as follows:

```python
class dict(MutableMapping[_KT, _VT]):
    ...
    def __ror__(self, __value: Mapping[_T1, _T2]) -> dict[_KT | _T1, _VT | _T2]: ...
```

This means the type of the operation is `dict[str, A | B]`, and this isn't assignable to a variable of type `dict[str, A]` because the type parameters for `dict` are invariant.

Mypy appears to accept this, presumably because it treats `A | B` as equivalent to `A`. I'll need to think more about whether this is a legitimate assumption.

```python
class A: pass
class B(A): pass
def func(y: dict[str, A | B]) -> None:
    x: dict[str, A] = y # Mypy is OK with this
```

You can work around this issue by using a temporary variable:
```python
temp: dict[str, A] = {"a": B()}
merged = temp | overrides
```

This works because of [bidirectional type inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types).

**Comment by erictraut** [comment]
There was a bug here. Pyright contains logic that should evaluate `dict[str, A | B]` as equivalent to `dict[str, A]`, but this logic didn't correctly handle the opposite union ordering (`dict[str, B | A]`). That resulted in the false positive error.

This will be addressed in the next release.

Here's the [commit](https://github.com/microsoft/pyright/commit/c4624664850cb752e77ae1c858698ca57734c9bc).

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.269, which I just published. It will also be included in a future version of pylance.
